pkg/healthcheck/handler.go

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package healthcheck
+
+import (
+	"net/http"
+	"strconv"
+
+	"go.uber.org/zap"
+)
+
+// Handler provides a configurable health check handler
+type Handler struct {
+	Listener chan int
+	State    int
+	Logger   *zap.Logger
+	Port     int
+}
+
+// Serve starts a new handler, ready to serve requests
+func Serve(state int, port int, logger *zap.Logger) (Handler, error) {
+	h := Handler{Listener: make(chan int), State: state, Port: port, Logger: logger}
+	go h.Register()
+	return h, nil
+}
+
+// Register creates a new handler on a new server mux, based on the Handler's options
+func (h Handler) Register() {
+	// Start the health check server, used for reporting the liveness and readiness state for this component
+	h.Logger.Info("Health Check server started", zap.Int("http-port", h.Port))
+	healthCheckMux := http.NewServeMux()
+
+	healthCheckMux.HandleFunc("/", func(w http.ResponseWriter, _ *http.Request) {
+		w.WriteHeader(h.State)
+
+		// this is written only for response with an entity, so, it won't be used for a 204 - No content
+		w.Write([]byte("Server not available"))
+	})
+
+	hcPortStr := ":" + strconv.Itoa(h.Port)
+
+	go func() {
+		if err := http.ListenAndServe(hcPortStr, healthCheckMux); err != nil {
```

Comment:
this file should have tests. Once you start doing tests you will find out that you want a Stop or Close method on the Handler. You can see how that's done in the agent's HTTP server.